File: zeroinstall/gtkui/trust_box.py
"""
A dialog box for confirming GPG keys.
"""

# Copyright (C) 2009, Thomas Leonard
# -*- coding: utf-8 -*-
# See the README file for details, or visit http://0install.net.
from zeroinstall import _

import gtk
from zeroinstall.injector.model import SafeException
from zeroinstall.injector import gpg, trust
from zeroinstall.support import tasks
from zeroinstall.gtkui import help_box, gtkutils

def frame(page, title, content, expand = False):
	frame = gtk.Frame()
	label = gtk.Label()
	label.set_markup('<b>%s</b>' % title)
	frame.set_label_widget(label)
	frame.set_shadow_type(gtk.SHADOW_NONE)
	if type(content) in (str, unicode):
		content = gtk.Label(content)
		content.set_alignment(0, 0.5)
		content.set_selectable(True)
	frame.add(content)
	if hasattr(content, 'set_padding'):
		content.set_padding(8, 4)
	else:
		content.set_border_width(8)
	page.pack_start(frame, expand, True, 0)

def pretty_fp(fp):
	s = fp[0:4]
	for x in range(4, len(fp), 4):
		s += ' ' + fp[x:x + 4]
	return s

def left(text):
	label = gtk.Label(text)
	label.set_alignment(0, 0.5)
	label.set_selectable(True)
	return label

def make_hints_area(closed, key_info_fetcher):
	def text(parent):
		text = ""
		for node in parent.childNodes:
			if node.nodeType == node.TEXT_NODE:
				text = text + node.data
		return text

	hints = gtk.VBox(False, 4)

	shown = set()
	def add_hints():
		infos = set(key_info_fetcher.info) - shown
		for info in infos:
			hints.add(make_hint(info.getAttribute("vote"), text(info)))
			shown.add(info)

		if not(key_info_fetcher.blocker or shown):
			hints.add(make_hint("bad", _('Warning: Nothing known about this key!')))

	if key_info_fetcher.blocker:
		status = left(key_info_fetcher.status)
		hints.add(status)

		@tasks.async
		def update_when_ready():
			while key_info_fetcher.blocker:
				yield key_info_fetcher.blocker, closed
				if closed.happened:
					# The dialog box was closed. Stop updating.
					return
				add_hints()
			status.destroy()
		update_when_ready()
	else:
		add_hints()

	hints.show()
	return hints

def make_hint(vote, hint_text):
	hint_icon = gtk.Image()
	if vote == "good":
		hint_icon.set_from_stock(gtk.STOCK_YES, gtk.ICON_SIZE_BUTTON)
	else:
		hint_icon.set_from_stock(gtk.STOCK_DIALOG_WARNING, gtk.ICON_SIZE_BUTTON)
	hint = left(hint_text)
	hint.set_line_wrap(True)
	hint_hbox = gtk.HBox(False, 4)
	hint_hbox.pack_start(hint_icon, False, True, 0)
	hint_hbox.pack_start(hint, True, True, 0)
	hint_icon.set_alignment(0, 0)
	hint_hbox.show_all()
	return hint_hbox

class TrustBox(gtk.Dialog):
	"""Display a dialog box asking the user to confirm that one of the
	keys is trusted for this domain.
	"""
	parent = None
	closed = None

	def __init__(self, pending, valid_sigs, parent):
		"""@since: 0.42"""
		assert valid_sigs

		gtk.Dialog.__init__(self)
		self.set_has_separator(False)
		self.set_position(gtk.WIN_POS_CENTER)
		self.set_transient_for(parent)

		self.closed = tasks.Blocker(_("confirming keys with user"))

		domain = trust.domain_from_url(pending.url)
		assert domain

		def destroy(box):
			self.closed.trigger()

		self.connect('destroy', destroy)

		self.set_title(_('Confirm trust'))

		vbox = gtk.VBox(False, 4)
		vbox.set_border_width(4)
		self.vbox.pack_start(vbox, True, True, 0)

		notebook = gtk.Notebook()

		if len(valid_sigs) == 1:
			notebook.set_show_tabs(False)

		label = left(_('Checking: %s') % pending.url)
		label.set_padding(4, 4)
		vbox.pack_start(label, False, True, 0)

		currently_trusted_keys = trust.trust_db.get_keys_for_domain(domain)
		if currently_trusted_keys:
			keys = [gpg.load_key(fingerprint) for fingerprint in currently_trusted_keys]
			descriptions = [_("%(key_name)s\n(fingerprint: %(key_fingerprint)s)") % {'key_name': key.name, 'key_fingerprint': pretty_fp(key.fingerprint)}
					for key in keys]
		else:
			descriptions = [_('None')]
		frame(vbox, _('Keys already approved for "%s"') % domain, '\n'.join(descriptions))

		label = left(ngettext('This key signed the feed:', 'These keys signed the feed:', len(valid_sigs)))

		label.set_padding(4, 4)
		vbox.pack_start(label, False, True, 0)

		vbox.pack_start(notebook, True, True, 0)

		self.add_button(gtk.STOCK_HELP, gtk.RESPONSE_HELP)
		self.add_button(gtk.STOCK_CANCEL, gtk.RESPONSE_CANCEL)
		self.add_button(gtk.STOCK_ADD, gtk.RESPONSE_OK)
		self.set_default_response(gtk.RESPONSE_OK)

		trust_checkbox = {}	# Sig -> CheckButton
		def ok_sensitive():
			trust_any = False
			for toggle in trust_checkbox.values():
				if toggle.get_active():
					trust_any = True
					break
			self.set_response_sensitive(gtk.RESPONSE_OK, trust_any)

		for sig in valid_sigs:
			if hasattr(sig, 'get_details'):
				name = '<unknown>'
				details = sig.get_details()
				for item in details:
					if item[0] == 'uid' and len(item) > 9:
						name = item[9]
						break
			else:
				name = None
			page = gtk.VBox(False, 4)
			page.set_border_width(8)

			frame(page, _('Fingerprint'), pretty_fp(sig.fingerprint))

			if name is not None:
				frame(page, _('Claimed identity'), name)

			frame(page, _('Unreliable hints database says'), make_hints_area(self.closed, valid_sigs[sig]))

			already_trusted = trust.trust_db.get_trust_domains(sig.fingerprint)
			if already_trusted:
				frame(page, _('You already trust this key for these domains'),
					'\n'.join(already_trusted))

			trust_checkbox[sig] = gtk.CheckButton(_('_Trust this key'))
			page.pack_start(trust_checkbox[sig], False, True, 0)
			trust_checkbox[sig].connect('toggled', lambda t: ok_sensitive())

			notebook.append_page(page, gtk.Label(name or 'Signature'))

		ok_sensitive()
		self.vbox.show_all()

		def response(box, resp):
			if resp == gtk.RESPONSE_HELP:
				trust_help.display()
				return
			if resp == gtk.RESPONSE_OK:
				self.trust_keys([sig for sig in trust_checkbox if trust_checkbox[sig].get_active()], domain)
			self.destroy()
		self.connect('response', response)
	
	def trust_keys(self, agreed_sigs, domain):
		assert domain
		try:
			for sig in agreed_sigs:
				trust.trust_db.trust_key(sig.fingerprint, domain)

			trust.trust_db.notify()
		except Exception, ex:
			gtkutils.show_message_box(self, str(ex), gtk.MESSAGE_ERROR)
			if not isinstance(ex, SafeException):
				raise

trust_help = help_box.HelpBox(_("Trust Help"),
(_('Overview'), '\n' +
_("""When you run a program, it typically has access to all your files and can generally do \
anything that you're allowed to do (delete files, send emails, etc). So it's important \
to make sure that you don't run anything malicious.""")),

(_('Digital signatures'), '\n' +
_("""Each software author creates a 'key-pair'; a 'public key' and a 'private key'. Without going \
into the maths, only something encrypted with the private key will decrypt with the public key.

So, when a programmer releases some software, they encrypt it with their private key (which no-one \
else has). When you download it, the injector checks that it decrypts using their public key, thus \
proving that it came from them and hasn't been tampered with.""")),

(_('Trust'), '\n' +
_("""After the injector has checked that the software hasn't been modified since it was signed with \
the private key, you still have the following problems:

1. Does the public key you have really belong to the author?
2. Even if the software really did come from that person, do you trust them?""")),

(_('Key fingerprints'), '\n' +
_("""To confirm (1), you should compare the public key you have with the genuine one. To make this \
easier, the injector displays a 'fingerprint' for the key. Look in mailing list postings or some \
other source to check that the fingerprint is right (a different key will have a different \
fingerprint).

You're trying to protect against the situation where an attacker breaks into a web site \
and puts up malicious software, signed with the attacker's private key, and puts up the \
attacker's public key too. If you've downloaded this software before, you \
should be suspicious that you're being asked to confirm another key!""")),

(_('Reputation'), '\n' +
_("""In general, most problems seem to come from malicous and otherwise-unknown people \
replacing software with modified versions, or creating new programs intended only to \
cause damage. So, check your programs are signed by a key with a good reputation!""")))

File: src/gui_gtk/trust_box.ml
(* Copyright (C) 2013, Thomas Leonard
 * See the README file for details, or visit http://0install.net.
 *)

open Support
open Support.Common
open Gtk_common

module Progress = Zeroinstall.Progress
module Trust = Zeroinstall.Trust
module G = Support.Gpg
module U = Support.Utils

let left text =
  GMisc.label ~text ~selectable:true ~xalign:0.0 ~yalign:0.5

let pretty_fp fingerprint =
  let b = Buffer.create (String.length fingerprint * 2) in
  let chunks = String.length fingerprint / 4 in
  for i = 0 to chunks - 1 do
    if i > 0 then
      Buffer.add_char b ' ';
    Buffer.add_substring b fingerprint (i * 4) 4;
  done;
  let tail = chunks * 4 in
  Buffer.add_substring b fingerprint tail (String.length fingerprint - tail);
  Buffer.contents b

let make_hint vote hint_text =
  let stock =
    match vote with
    | Progress.Good -> `YES
    | Progress.Bad -> `DIALOG_WARNING in
  let hint_icon = GMisc.image ~stock ~icon_size:`BUTTON ~xalign:0.0 ~yalign:0.0 () in
  let hint = left hint_text ~line_wrap:true () in
  let hint_hbox = GPack.hbox ~homogeneous:false ~spacing:4 () in
  hint_hbox#pack (hint_icon :> GObj.widget) ~expand:false ~fill:true;
  hint_hbox#pack (hint :> GObj.widget) ~expand:true ~fill:true;
  hint_hbox

let make_hints_area result hints =
  let box = GPack.vbox ~border_width:8 ~homogeneous:false ~spacing:4 () in

  let add_hints hints =
    if Lwt.state result = Lwt.Sleep then (
      let hints =
        if hints = [] then [(Progress.Bad, "Warning: Nothing known about this key!")] else hints in
      hints |> List.iter (fun (vote, msg) ->
        box#pack ~expand:false ~fill:true (make_hint vote msg :> GObj.widget)
      )
    ) in

  begin match Lwt.state hints with
  | Lwt.Sleep ->
      let label = left "Waiting for response from key information server..." () in
      box#pack ~expand:false ~fill:true (label :> GObj.widget);
      Gtk_utils.async (fun () ->
        hints >|= fun hints ->
        label#destroy ();
        add_hints hints
      )
  | Lwt.Return hints -> add_hints hints
  | Lwt.Fail ex -> raise ex end;

  box

let confirm_keys_help = Help_box.create "Trust Help" [
("Overview",
"When you run a program, it typically has access to all your files and can generally do \
anything that you're allowed to do (delete files, send emails, etc). So it's important \
to make sure that you don't run anything malicious.");

("Digital signatures",
"Each software author creates a 'key-pair'; a 'public key' and a 'private key'.\
\n\n\
When a programmer releases some software, they sign it with their private key (which no-one \
else has). When you download it, 0install checks the signature using the public key, thus \
proving that it came from them and hasn't been tampered with.");

("Trust",
"After 0install has checked that the software hasn't been modified since it was signed with \
the private key, you still have the following problems:\
\n\n\
1. Does the public key you have really belong to the author?\n\
2. Even if the software really did come from that person, do you trust them?");

("Key fingerprints",
"To confirm (1), you should compare the public key you have with the genuine one. To make this \
easier, 0install displays a 'fingerprint' for the key. Look in mailing list postings or some \
other source to check that the fingerprint is right (a different key will have a different \
fingerprint).\
\n\n\
You're trying to protect against the situation where an attacker breaks into a web site \
and puts up malicious software, signed with the attacker's private key, and puts up the \
attacker's public key too. If you've downloaded this software before, you \
should be suspicious that you're being asked to confirm another key!");

("Reputation",
"In general, most problems seem to come from malicous and otherwise-unknown people \
replacing software with modified versions, or creating new programs intended only to \
cause damage. So, check your programs are signed by a key with a good reputation!");
]

(** Check the key-info server's results for these keys. If we don't know any of them,
 * ask for extra confirmation from the user.
 * @return true if we should continue and trust the keys. *)
let confirm_unknown_keys ~parent to_trust valid_sigs =
  let unknown = to_trust |> List.filter (fun fpr ->
    let hints = List.assoc fpr valid_sigs in
    match Lwt.state hints with
    | Lwt.Fail _ | Lwt.Sleep -> true  (* Still waiting => unknown *)
    | Lwt.Return hints ->
        (* Unknown if we didn't get any Good votes *)
        not (List.exists (fun (vote, _msg) -> vote = Progress.Good) hints)
  ) in
  let confirm message =
    let result, set_result = Lwt.wait () in
    let box = GWindow.message_dialog
      ~message
      ~message_type:`QUESTION
      ~parent
      ~destroy_with_parent:true
      ~buttons:GWindow.Buttons.ok_cancel
      ~position:`CENTER
      () in
    box#action_area#set_spacing 4;
    box#action_area#set_border_width 4;
    box#connect#response ==> (function
      | `OK -> Lwt.wakeup set_result true; box#destroy ()
      | `DELETE_EVENT | `CANCEL -> Lwt.wakeup set_result false; box#destroy ()
    );
    box#show ();
    result in
  match unknown with
  | [] -> Lwt.return true
  | [_] -> confirm "WARNING: you are confirming a key which was not known to the key server. Are you sure?"
  | _   -> confirm "WARNING: you are confirming keys which were not known to the key server. Are you sure?"

let frame ~title ~content (parent:GPack.box) =
  let frame = GBin.frame ~shadow_type:`NONE () in
  let label = GMisc.label
    ~markup:(Printf.sprintf "<b>%s</b>" title)    (* Escaping? *)
    () in
  frame#set_label_widget (Some (label :> GObj.widget));
  frame#add (content :> GObj.widget);
  parent#pack (frame :> GObj.widget) ~expand:false ~fill:true

let confirm_keys gpg trust_db ?parent feed_url valid_sigs =
  assert (valid_sigs <> []);
  let n_sigs = List.length valid_sigs in
  let `Remote_feed url = feed_url in

  valid_sigs |> List.map fst |> G.load_keys gpg >>= fun key_names ->

  let result, set_result = Lwt.wait () in
  let dialog = GWindow.dialog
    ?parent
    ~title:"Confirm trust"
    ~position:`CENTER
    () in
  dialog#action_area#set_border_width 4;

  let vbox = GPack.vbox
    ~homogeneous:false
    ~spacing:4
    ~border_width:4
    () in

  dialog#vbox#pack ~expand:true ~fill:true (vbox :> GObj.widget);

  let label = left (Printf.sprintf "Checking: %s" url) ~xpad:4 ~ypad:4 () in
  vbox#pack ~expand:false ~fill:true (label :> GObj.widget);

  let domain = Trust.domain_from_url feed_url in

  begin match trust_db#get_keys_for_domain domain |> XString.Set.elements with
    | [] -> Lwt.return ["None"]
    | keys ->
      G.load_keys gpg keys >|= XString.Map.map_bindings
        (fun fp info ->
          Printf.sprintf "%s\n(fingerprint: %s)" (default "?" info.G.name) (pretty_fp fp)
        )
  end >>= fun descriptions ->

  frame
    ~title:(Printf.sprintf "Keys already approved for '%s'" domain)
    ~content:(left (String.concat "\n" descriptions) ~xpad:8 ~ypad:4 ())
    vbox;

  let label =
    match valid_sigs with
    | [_] -> "This key signed the feed:"
    | _ -> "These keys signed the feed:" in
  vbox#pack ~expand:false ~fill:true (left label ~xpad:4 ~ypad:4 () :> GObj.widget);

  let notebook = GPack.notebook
    ~show_border:false
    ~show_tabs:(n_sigs > 1)
    () in

  dialog#add_button_stock `HELP `HELP;
  (* Lablgtk uses the wrong response code for HELP, so we have to do this manually. *)
  let actions = dialog#action_area in
  actions#set_child_secondary (List.hd actions#children) true;

  dialog#add_button_stock `CANCEL `CANCEL;
  dialog#add_button_stock `ADD `OK;
  dialog#set_default_response `OK;

  let trust_checkboxes = ref [] in

  (* The OK button is available whenever at least one key is selected to be trusted. *)
  let update_ok_button () =
    !trust_checkboxes
    |> List.exists (fun (_fpr, box) -> box#active)
    |> dialog#set_response_sensitive `OK in

  valid_sigs |> List.iter (fun (fpr, hints) ->
    let name =
      XString.Map.find_opt fpr key_names
      |> pipe_some (fun info -> info.G.name)
      |> default "<unknown>" in
    let page = GPack.vbox ~homogeneous:false ~spacing:4 ~border_width:8 () in
    frame
      ~title:"Fingerprint"
      ~content:(left (pretty_fp fpr) ~xpad:8 ~ypad:4 ())
      page;
    frame
      ~title:"Claimed identity"
      ~content:(left name ~xpad:8 ~ypad:4 ())
      page;
    frame
      ~title:"Unreliable hints database says"
      ~content:(make_hints_area result hints)
      page;

    let already_trusted = trust_db#get_trust_domains fpr |> XString.Set.elements in
    if already_trusted <> [] then (
      frame
        ~title:"You already trust this key for these domains"
        ~content:(left (String.concat "\n" already_trusted) ~xpad:8 ~ypad:4 ())
        page
    );

    let checkbox = GButton.check_button
      ~use_mnemonic:true
      ~show:(n_sigs > 1)
      ~label:"_Trust this key" () in
    trust_checkboxes := (fpr, checkbox) :: !trust_checkboxes;
    page#pack ~expand:false ~fill:true (checkbox :> GObj.widget);
    checkbox#connect#toggled ==> (fun _cb -> update_ok_button ());
    let tab_label = (GMisc.label ~text:name () :> GObj.widget) in
    append_page notebook ~tab_label (page :> GObj.widget);
  );
  (List.hd !trust_checkboxes |> snd)#set_active true;

  vbox#pack ~expand:true ~fill:true (notebook :> GObj.widget);

  dialog#connect#response ==> (function
    | `OK ->
        let to_trust = !trust_checkboxes |> List.filter_map (fun (fpr, box) ->
          if box#active then Some fpr else None
        ) in
        assert (to_trust <> []);
        let ok = confirm_unknown_keys ~parent:dialog to_trust valid_sigs in
        Gtk_utils.async ~parent:dialog (fun () ->
          ok >|= function
          | false -> ()
          | true ->
              Lwt.wakeup set_result to_trust;
              dialog#destroy ()
        )
    | `DELETE_EVENT | `CANCEL -> Lwt.wakeup set_result []; dialog#destroy ()
    | `HELP -> confirm_keys_help#display
  );

  dialog#show ();
  result
